common/kafka/kafka_consumer_holder.cpp

```diff
@@ -0,0 +1,448 @@
+    LOG(INFO) << "Recreating kafka consumer for topic: "
+              << folly::join(",", topic_names_)
+              << ", partitions: "
+              << folly::join(",", partition_ids_)
+              << std::endl << std::flush;
+
+    kafkaConsumer_ = CreateRdKafkaConsumer(
+      partition_ids_,
+      broker_list_,
+      group_id_,
+      kafka_consumer_type_);
+
+    LOG(INFO) << "Recreated kafka consumer for topic: "
+              << folly::join(",", topic_names_)
+              << ", partitions: "
+              << folly::join(",", partition_ids_)
+              << std::endl << std::flush;
+  }
+
+  bool assignPartitions() {
+    /**
+     * First assign the topic partitions to current kafka consumer.
+     */
+    // A temporary RdKafka::TopicPartition* vector specifically used
+    // for KafkaConsumer->assign(). The pointers inside this vector is
+    // owned by `topic_partitions`.
+
+    LOG(INFO) << "Assigning kafka consumer for topic: "
+              << folly::join(",", topic_names_)
+              << ", partitions: "
+              << folly::join(",", partition_ids_)
+              << std::endl << std::flush;
+
+    std::vector<std::shared_ptr<RdKafka::TopicPartition>> topic_partitions;
+    std::vector<RdKafka::TopicPartition*> tmp_topic_partitions;
+    topic_partitions.reserve(partition_ids_.size() * topic_names_.size());
+    tmp_topic_partitions.reserve(partition_ids_.size() * topic_names_.size());
+    for (const auto partition_id : partition_ids_) {
+      for (const auto& topic_name : topic_names_) {
+        topic_partitions.push_back(
+          std::shared_ptr<RdKafka::TopicPartition>(
+            RdKafka::TopicPartition::create(
+              topic_name,
+              partition_id,
+              // Use the end of the offset to init here, but it does not matter.
+              // When kafkaconsumer is used, the caller is supposed to call
+              // seek() to the expected offset before consuming messages.
+              RdKafka::Topic::OFFSET_END)));
+        tmp_topic_partitions.push_back(topic_partitions.back().get());
+      }
+    }
+    const auto error_code = ExecuteKafkaOperationWithRetry(
+      [this, &tmp_topic_partitions]() {
+        return kafkaConsumer_->assign(tmp_topic_partitions);
+      });
+
+    if (error_code != RdKafka::ERR_NO_ERROR) {
+      LOG(INFO) << "Failed to assign kafka consumer for topic: "
+                << folly::join(",", topic_names_)
+                << ", partitions: "
+                << folly::join(",", partition_ids_)
+                << std::endl << std::flush;
+
+      return false;
+    }
+
+    LOG(INFO) << "Assigned kafka consumer for topic: "
+              << folly::join(",", topic_names_)
+              << ", partitions: "
+              << folly::join(",", partition_ids_)
+              << std::endl << std::flush;
+
+    return true;
+  }
+
+  bool seekToRelevantOffsets(const TopicPartitionToValueMap<ConsumedOffset>* consumedOffsets) {
```

Comment:
Is `consumedOffsets` an in-parameter? Can we pass it as a const reference?

Reply:
Sure.